File: nexus-ai-operators/app/engine/dead_letter.py
import json
from datetime import datetime, timezone
from pathlib import Path
from threading import Lock


class DeadLetterLog:
    def __init__(self, path: str = "data/dead_letters.jsonl") -> None:
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self._lock = Lock()

    def append(self, event_id: str, payload: dict, error: str, attempts: int) -> None:
        entry = {
            "event_id": event_id,
            "payload": payload,
            "error": error,
            "attempts": attempts,
            "failed_at": datetime.now(timezone.utc).isoformat(),
        }
        with self._lock:
            with open(self.path, "a", encoding="utf-8") as f:
                f.write(json.dumps(entry) + "\n")

File: nexus-control-operators/internal/ops/eventstore/deadletter.go
package eventstore

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// DeadLetterEntry represents a permanently failed event persisted for replay/forensics.
type DeadLetterEntry struct {
	EventID  string    `json:"event_id"`
	Payload  any       `json:"payload"`
	Error    string    `json:"error"`
	Attempts int       `json:"attempts"`
	FailedAt time.Time `json:"failed_at"`
}

// DeadLetterLog stores failed events as JSONL.
type DeadLetterLog struct {
	mu   sync.Mutex
	path string
}

// NewDeadLetterLog creates a file-backed dead-letter logger.
func NewDeadLetterLog(path string) *DeadLetterLog {
	return &DeadLetterLog{path: path}
}

// Append writes one failed event line in JSONL format.
func (d *DeadLetterLog) Append(entry DeadLetterEntry) error {
	d.mu.Lock()
	defer d.mu.Unlock()

	if err := os.MkdirAll(filepath.Dir(d.path), 0o755); err != nil {
		return err
	}
	f, err := os.OpenFile(d.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	defer f.Close()

	line, err := json.Marshal(entry)
	if err != nil {
		return err
	}
	line = append(line, '\n')
	_, err = f.Write(line)
	return err
}
